### PR TITLE
Mtp baseline

### PR DESCRIPTION
**Pull request by freddyaboulton**
* Adds a baseline MTP model with Resnet, MobileNet backbones
* Converts models.py into a directory with physics.py and mtp.py
* Adds unit tests for mtp.py

How to test:
* Run unit tests `nosetests predict/tests` and `nosetests eval/predict/tests`
* The script `run_mtp.py` tests whether the mtp model can overfit on a single example (for 1 or two modes) with the cpu or gpu

@elenacorinagrigore Please verify the model implementation

@emwolff @holger-nutonomy your feedback is appreciated :smile:

### CODE REVIEW

File: python-sdk/nuscenes/eval/predict/config.py [hunk 1]
```diff
@@ -10,9 +10,11 @@ class PredictionConfig:
 
     def __init__(self,
                  metrics: List[Metric],
-                 seconds: int = 6):
+                 seconds: int = 6,
+                 frequency: int = 2):
         self.metrics = metrics
         self.seconds = seconds
+        self.frequency = 2
```
Comment (emwolff) [inline python-sdk/nuscenes/eval/predict/config.py]: Should be `self.frequency = frequency`, no?

Comment (emwolff) [reply python-sdk/nuscenes/eval/predict/config.py]: Should also add units (Hz) for frequency

File: python-sdk/nuscenes/predict/models/mtp.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+        """
+        backbone_features = self.backbone(input_tensor)
+        return backbone_features.mean([2, 3])
+
+
+class MTP(nn.Module):
+    """Implements the MTP network."""
+
+    def __init__(self, backbone: nn.Module, num_modes: int,
+                 seconds: int = 6, frequency_in_hz: int = 2,
```
Comment (emwolff) [inline python-sdk/nuscenes/predict/models/mtp.py]: Should `seconds` and `frequencey_in_hz` be `int`s?

File: python-sdk/nuscenes/predict/models/mtp.py [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+
+        self.fc2 = nn.Linear(n_hidden_layers, int(num_modes * predictions_per_mode + num_modes))
+
+    def _calculate_backbone_feature_dim(self, input_shape: Tuple[int, int, int]):
+        """Helper to calculate the shape of the fully-connected regression layer."""
+        tensor = torch.ones(1, *input_shape)
+        output_feat = self.backbone.forward(tensor)
+        return output_feat.shape[-1]
+
+    def forward(self, image_tensor, agent_state_vector):
```
Comment (emwolff) [inline python-sdk/nuscenes/predict/models/mtp.py]: Missing type hints.

File: python-sdk/nuscenes/predict/models/mtp.py [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+
+        return torch.cat((predictions, mode_probabilities), 1)
+
+
+class MTPLoss:
+    """
+    Computes the loss for the MTP model.
+    """
+
+    def __init__(self, num_modes, regression_loss_weight,
```
Comment (emwolff) [inline python-sdk/nuscenes/predict/models/mtp.py]: Missing type hints.

File: python-sdk/nuscenes/eval/predict/config.py (hunk 1 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/eval/predict/config.py]: Yes

File: python-sdk/nuscenes/predict/models/mtp.py (hunk 2 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/models/mtp.py]: Not necessarily but `seconds * frequency_in_hz` should be an int

File: python-sdk/nuscenes/predict/models/mtp.py [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        # We choose the best mode to be the one that provides the lowest ave of l2 norms between the
+        # predicted trajectory and the ground truth, taking into account only the modes with an angle
+        # less than the threshold IF there is at least one mode with an angle less than the threshold.
+        else:
+            # Out of the selected modes above, we choose the final best mode as that which returns the
+            # smallest ave of l2 norms between the predicted and ground truth trajectories.
+            distances_from_ground_truth = []
+
+            for angle, mode in angles_from_ground_truth[:max_angle_below_thresh_idx + 1]:
+                norm = self._compute_ave_l2_norms(target - trajectories[mode, :, :]).item()
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/predict/models/mtp.py]: `_compute_ave_l2_norms()` supposedly returns a float.
I suggest you move the `.item()` into that function.

File: python-sdk/nuscenes/predict/models/mtp.py [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+        :param frequency: Frequency between timesteps in the prediction (in HZ).
+            Highest frequency is nuScenes is 2 hz.
+        :param n_hidden_layers: Size of fully connected layer after the CNN
+            backbone processes the image.
+        :param input_shape: Shape of the input expected by the network.
+            This is needed because the size of the fully connected layer after
+            the backbone depends on the backbone and its version.
+
+        Note:
+            Although seconds and frequency_in_hz are labeled their
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/predict/models/mtp.py]: `are labeled as float` ?

File: python-sdk/nuscenes/predict/models/mtp.py [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+    def _calculate_backbone_feature_dim(self, input_shape: Tuple[int, int, int]) -> int:
+        """Helper to calculate the shape of the fully-connected regression layer."""
+        tensor = torch.ones(1, *input_shape)
+        output_feat = self.backbone.forward(tensor)
+        return output_feat.shape[-1]
+
+    def forward(self, image_tensor: torch.Tensor,
+                agent_state_vector: torch.Tensor) -> torch.Tensor:
+        """
+        Forward pass of the model.
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/predict/models/mtp.py]: Missing argument docstrings

File: python-sdk/nuscenes/predict/models/mtp.py [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+        return trajectories_no_modes, mode_probabilities
+
+    def _angle_between(self, ref_traj: torch.Tensor,
+                       traj_to_compare: torch.Tensor) -> float:
+        """
+        Computes the angle between the last points of the two trajectories.
+        The resulting angle is in degrees and is an angle in the [0; 180) interval.
+        :param ref_traj: Tensor of shape [n_timesteps, 2]
+        :param traj_to_compare: Tensor of shape [n_timesteps, 2]
+        """
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/predict/models/mtp.py]: Missing return

File: python-sdk/nuscenes/predict/models/mtp.py [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+
+    def _compute_best_mode(self,
+                           angles_from_ground_truth: List[float],
+                           target: torch.Tensor, trajectories: torch.Tensor) -> int:
+        """
+        Finds the index of the best mode given the angles from the ground truth.
+        :param angles_from_ground_truth: List of angles
+        :param target: Shape [1, n_timesteps, 2]
+        :param trajectories: Shape [n_modes, n_timesteps, 2]
+        """
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/predict/models/mtp.py]: returns

File: python-sdk/nuscenes/predict/models/mtp.py [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+
+                distances_from_ground_truth.append((norm, mode))
+
+            distances_from_ground_truth = sorted(distances_from_ground_truth)
+            best_mode = distances_from_ground_truth[0][1]
+
+        return best_mode
+
+    def __call__(self, predictions: torch.Tensor, targets: torch.Tensor) -> torch.Tensor:
+
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/predict/models/mtp.py]: Could use some comments

File: python-sdk/nuscenes/predict/tests/run_mtp.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    current_loss = 10000
+
+    optimizer = optim.SGD(model.parameters(), lr=0.1)
+
+    n_iter = 0
+
+    minimum_loss = 0
+
+    if args.num_modes == 2:
+        minimum_loss += 0.56234
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/predict/tests/run_mtp.py]: Where does this number come from?

File: python-sdk/nuscenes/predict/tests/run_mtp.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
+import argparse
+import torch
+import numpy as np
+from torch.utils.data import DataLoader, IterableDataset
+import torch.optim as optim
+
+from nuscenes.predict.models.mtp import ResNetBackbone, MTP, MTPLoss
+
+class TestDataset(IterableDataset):
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/predict/tests/run_mtp.py]: Let's add a comment for what this test does

File: python-sdk/nuscenes/predict/tests/test_mtp.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+from torchvision.models.resnet import BasicBlock, Bottleneck
+from nuscenes.predict.models import mtp
+
+class TestBackBones(unittest.TestCase):
+
+    def count_layers(self, model):
+        if isinstance(model[4][0], BasicBlock):
+            n_convs = 2
+        elif isinstance(model[4][0], Bottleneck):
+            n_convs = 3
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/predict/tests/test_mtp.py]: Suggest to have an else case here in case none of the above is true.

File: python-sdk/nuscenes/predict/tests/test_mtp_loss.py [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+
+    def test_get_trajectories_and_modes(self,):
+
+        loss_n_modes_5 = mtp.MTPLoss(5, 0, 0)
+        loss_n_modes_1 = mtp.MTPLoss(1, 0, 0)
+
+        xy_pred = torch.arange(60).view(1, -1).repeat(1, 5).view(-1, 60)
+        mode_pred = torch.arange(5).view(1, -1)
+
+        prediction_bs_1 = torch.cat([xy_pred.reshape(1, -1), mode_pred], axis=1)
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/predict/tests/test_mtp_loss.py]: axis or dim ?

File: python-sdk/nuscenes/predict/tests/test_mtp_loss.py [hunk 15]
```diff
@@ -0,0 +1,193 @@
+
+import unittest
+import torch
+from nuscenes.predict.models import mtp
+import math
+
+class TestMTPLoss(unittest.TestCase):
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/predict/tests/test_mtp_loss.py]: Add a comment

File: python-sdk/nuscenes/predict/tests/test_mtp_loss.py [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+
+    def test_angle_between_trajectories(self,):
+
+        def make_trajectory(last_point):
+            traj = torch.zeros((12, 2))
+            traj[-1] = torch.Tensor(last_point)
+            return traj
+
+        reference = torch.zeros(12, 2)
+        trajectory = torch.zeros(12, 2)
```
Comment (holger-nutonomy) [inline python-sdk/nuscenes/predict/tests/test_mtp_loss.py]: reference and trajectory are unused

File: python-sdk/nuscenes/predict/models/mtp.py [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+from torch.nn import functional as f
+from torchvision.models import (mobilenet_v2, resnet18, resnet34, resnet50,
+                                resnet101, resnet152)
+
+from nuscenes.eval.predict.config import PredictionConfig
+
+
+def trim_network_at_index(network: nn.Module, index: int = -1) -> nn.Module:
+    """
+    Returns a new network with all layers up to index from the back.
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/models/mtp.py]: Added an assert to ensure the index is negative. Good catch!

File: python-sdk/nuscenes/predict/models/mtp.py [hunk 18]
```diff
@@ -0,0 +1,325 @@
...
+
+    def forward(self, image_tensor: torch.Tensor,
+                agent_state_vector: torch.Tensor) -> torch.Tensor:
+        """
+        Forward pass of the model.
+        """
+
+        backbone_features = self.backbone(image_tensor)
+
+        features = torch.cat([backbone_features, agent_state_vector], axis=1)
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/models/mtp.py]: Changed! Although I'm confused as to what it does if you pass in `axis`.

File: python-sdk/nuscenes/predict/models/mtp.py (hunk 6 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/models/mtp.py]: `typed` :relaxed:

File: python-sdk/nuscenes/predict/models/mtp.py (hunk 5 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/models/mtp.py]: Good idea.

File: python-sdk/nuscenes/predict/models/mtp.py [hunk 19]
```diff
@@ -0,0 +1,325 @@
...
+            Second item are the mode probabilities of shape [batch_size, num_modes]
+        """
+        mode_probabilities = model_prediction[:, -self.num_modes:].clone()
+
+        desired_shape = (model_prediction.shape[0], self.num_modes, -1, self.num_location_coordinates_predicted)
+        trajectories_no_modes = model_prediction[:, :-self.num_modes].clone().reshape(desired_shape)
+
+        return trajectories_no_modes, mode_probabilities
+
+    def _angle_between(self, ref_traj: torch.Tensor,
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/models/mtp.py]: Done!

File: python-sdk/nuscenes/predict/models/mtp.py (hunk 7 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/models/mtp.py]: Added.

File: python-sdk/nuscenes/predict/models/mtp.py (hunk 8 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/models/mtp.py]: Added

File: python-sdk/nuscenes/predict/models/mtp.py (hunk 9 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/models/mtp.py]: Done

File: python-sdk/nuscenes/predict/models/mtp.py (hunk 10 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/models/mtp.py]: Done!

File: python-sdk/nuscenes/predict/models/mtp.py [hunk 20]
```diff
@@ -0,0 +1,328 @@
...
+
+class MTPLoss:
+    """
+    Computes the loss for the MTP model.
+    """
+
+    def __init__(self,
+                 num_modes: int,
+                 regression_loss_weight: float,
+                 angle_threshold_degrees: float = 5.):
```
Comment (elenacorinagrigore) [inline python-sdk/nuscenes/predict/models/mtp.py]: @freddyaboulton We've been using `7` as the default value for the `angle_threshold_degrees` param for our models.

File: python-sdk/nuscenes/predict/models/mtp.py [hunk 21]
```diff
@@ -0,0 +1,328 @@
...
+        Inits MTP loss
+        :param num_modes: How many modes are being predicted for each agent.
+        :param regression_loss_weight: Coefficient applied to the regression loss to
+            balance classification and regression performance.
+        :param angle_threshold_degrees: Minimum angle needed between a predicted trajectory
+            and the ground to consider it a match.
+        """
+        self.num_modes = num_modes
+        self.num_location_coordinates_predicted = 2  # We predict x, y coordinates at each timestep.
+        self.regression_loss_weight = regression_loss_weight
```
Comment (elenacorinagrigore) [inline python-sdk/nuscenes/predict/models/mtp.py]: @freddyaboulton We've been using `1` as the default for the `regression_loss_weight` param for our models.

File: python-sdk/nuscenes/predict/tests/run_mtp.py (hunk 12 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/tests/run_mtp.py]: I've changed `TestDataset` to `Dataset` and added a module docstring explaining what this script does.

File: python-sdk/nuscenes/predict/tests/test_mtp.py (hunk 13 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/tests/test_mtp.py]: Done!

File: python-sdk/nuscenes/predict/tests/test_mtp_loss.py (hunk 14 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/tests/test_mtp_loss.py]: Changed to dim!

File: python-sdk/nuscenes/predict/tests/test_mtp_loss.py (hunk 15 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/tests/test_mtp_loss.py]: Done!

File: python-sdk/nuscenes/predict/tests/test_mtp_loss.py (hunk 16 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/tests/test_mtp_loss.py]: Removed!

File: python-sdk/nuscenes/predict/tests/run_mtp.py (hunk 11 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/tests/run_mtp.py]: We expect to see 75% positive and 25% negative example so the minimum regression loss is expected to be 0.562. I'll add a comment explaining this.

File: python-sdk/nuscenes/predict/models/mtp.py [hunk 22]
```diff
@@ -0,0 +1,328 @@
...
+            Although seconds and frequency_in_hz are labeled their
+            product should be an int.
+        """
+
+        super().__init__()
+
+        self.backbone = backbone
+        self.num_modes = num_modes
+        backbone_feature_dim = self._calculate_backbone_feature_dim(input_shape)
+        self.fc1 = nn.Linear(backbone_feature_dim + 3, n_hidden_layers)
```
[the remark this answers is not in the document]
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/models/mtp.py]: ASV dim.

File: python-sdk/nuscenes/predict/models/mtp.py (hunk 21 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/models/mtp.py]: Adding now!

File: python-sdk/nuscenes/predict/models/mtp.py (hunk 20 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/predict/models/mtp.py]: Thanks! Think I'll keep it at five since that's what they mentioned in the paper.

**Comment by freddyaboulton** [comment]
@elenacorinagrigore Thanks for the review!

File: python-sdk/nuscenes/predict/models/mtp.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (holger-nutonomy) [reply python-sdk/nuscenes/predict/models/mtp.py]: Me too :-D. Maybe it's still supported for backward-compatibility.
